lib/features/campaigns/helper/campaign_action_cache.dart

```diff
@@ -0,0 +1,453 @@
+          var model = action.getAsDoorUpdate();
+          markerItems.add(model.transformToVirtualMarkerItem());
+        case CampaignActionType.deleteDoor:
+          var model = _getDeleteMarkerModel(PoiServiceType.door, action.poiId!);
+          markerItems.add(model);
+
+        case CampaignActionType.addFlyer:
+          var model = action.getAsFlyerCreate();
+          markerItems.add(model.transformToVirtualMarkerItem(action.poiTempId));
+        case CampaignActionType.editFlyer:
+          var model = action.getAsFlyerUpdate();
+          markerItems.add(model.transformToVirtualMarkerItem());
+        case CampaignActionType.deleteFlyer:
+          var model = _getDeleteMarkerModel(PoiServiceType.flyer, action.poiId!);
+          markerItems.add(model);
+
+        case CampaignActionType.unknown:
+        case null:
+          throw UnimplementedError();
+      }
+    }
+    return markerItems;
+  }
+
+  Future<PosterDetailModel> getPoiAsPosterDetail(String poiId) async {
+    var detailModel = await _getPoiDetail<PosterDetailModel>(
+      poiId: poiId,
+      addActionFilter: CampaignActionType.addPoster,
+      editActionFilter: CampaignActionType.editPoster,
+      transformEditAction: (action) => action.getAsPosterUpdate().transformToPosterDetailModel(),
+      transformAddAction: (action) => action.getAsPosterCreate().transformToPosterDetailModel(poiId),
+    );
+    return detailModel;
+  }
+
+  Future<DoorDetailModel> getPoiAsDoorDetail(String poiId) async {
+    var detailModel = await _getPoiDetail<DoorDetailModel>(
+      poiId: poiId,
+      addActionFilter: CampaignActionType.addDoor,
+      editActionFilter: CampaignActionType.editDoor,
+      transformEditAction: (action) => action.getAsDoorUpdate().transformToDoorDetailModel(),
+      transformAddAction: (action) => action.getAsDoorCreate().transformToDoorDetailModel(poiId),
+    );
+    return detailModel;
+  }
+
+  Future<FlyerDetailModel> getPoiAsFlyerDetail(String poiId) async {
+    var detailModel = await _getPoiDetail<FlyerDetailModel>(
+      poiId: poiId,
+      addActionFilter: CampaignActionType.addFlyer,
+      editActionFilter: CampaignActionType.editFlyer,
+      transformEditAction: (action) => action.getAsFlyerUpdate().transformToFlyerDetailModel(),
+      transformAddAction: (action) => action.getAsFlyerCreate().transformToFlyerDetailModel(poiId),
+    );
+    return detailModel;
+  }
+
+  Future<T> _getPoiDetail<T>({
+    required String poiId,
+    required CampaignActionType addActionFilter,
+    required CampaignActionType editActionFilter,
+    required T Function(CampaignAction) transformEditAction,
+    required T Function(CampaignAction) transformAddAction,
+  }) async {
+    var cacheList = await _findActionsByPoiId(poiId);
+    var addActions = cacheList.where((p) => p.actionType == addActionFilter).toList();
+    var editActions = cacheList.where((p) => p.actionType == editActionFilter).toList();
+    if (editActions.isNotEmpty) {
```

Comment:
This probably has minimal real-world impact, but technically the section could be refactored a tiny bit to save one op (I think just accepting my suggestion won't work because of the github line insertion point):
```suggestion
    var cacheList = await _findActionsByPoiId(poiId);
    var editActions = cacheList.where((p) => p.actionType == editActionFilter).toList();
    if (editActions.isNotEmpty) {
      var editAction = editActions.single;
      return transformEditAction(editAction);
    } else {
      var addActions = cacheList.where((p) => p.actionType == addActionFilter).toList();
      var addAction = addActions.single;
      return transformAddAction(addAction);
    }
```
``